pkg/jobs/hive/hive.go

```diff
@@ -181,3 +187,189 @@ func monitorDeployStatus(client clientv1.Client, hiveset hiveclient.Interface, c
 	}
 	return nil
 }
+
+func UpgradeCluster(client clientv1.Client, clusterName string, curator *clustercuratorv1.ClusterCurator) error {
+	klog.V(0).Info("* Initiate Upgrade")
+	klog.V(2).Info("Looking up managedclusterinfo " + clusterName)
+
+	desiredUpdate := curator.Spec.Upgrade.DesiredUpdate
+
+	managedClusterInfo := managedclusterinfov1beta1.ManagedClusterInfo{}
+	if err := client.Get(context.TODO(), types.NamespacedName{
+		Namespace: clusterName,
+		Name:      clusterName,
+	}, &managedClusterInfo); err != nil {
+		return err
+	}
+
+	klog.V(2).Info("kubevendor ", managedClusterInfo.Status.KubeVendor)
+
+	if managedClusterInfo.Status.KubeVendor != "OpenShift" {
+		return errors.New("Can not upgrade non openshift cluster")
+	}
+
+	if desiredUpdate == "" {
+		return errors.New("Provide valid upgrade version")
+	}
+
+	isValidVersion := false
+	if managedClusterInfo.Status.DistributionInfo.OCP.AvailableUpdates != nil {
+		for _, version := range managedClusterInfo.Status.DistributionInfo.OCP.AvailableUpdates {
+			if version == desiredUpdate {
+				isValidVersion = true
+			}
+		}
+	}
+	if !isValidVersion {
+		return errors.New("Provided version is not valid")
+	}
+
+	klog.V(2).Info("Create managedclusterview " + clusterName)
+
+	managedclusterview := &managedclusterviewv1beta1.ManagedClusterView{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      clusterName,
+			Namespace: clusterName,
+		},
+		Spec: managedclusterviewv1beta1.ViewSpec{
+			Scope: managedclusterviewv1beta1.ViewScope{
+				Group:     "config.openshift.io",
+				Kind:      "ClusterVersion",
+				Name:      "version",
+				Namespace: "",
+				Version:   "v1",
+			},
+		},
+	}
+
+	if err := client.Create(context.TODO(), managedclusterview); err != nil {
+		return err
+	}
+	resultmcview := managedclusterviewv1beta1.ManagedClusterView{}
+	for i := 1; i <= 5; i++ {
```

Comment:
is 5 times long enough? and I didn't see a timeout path (e.g. when raw is nil what will happen?)

Reply:
Actually @hanqiuzh I think this loop only runs 1 iteration. 
1. Pause 5s
2. check for managedClusterView
3. If found continue, if not found you get an "err" about not being found and return err... 

So maybe the "return" should be "continue" with a log of the "err" and a timeout when i=4 at Ln 255 that does a return.

		}, &resultmcview); err != nil {
			if i == 4 {
			  return err
			}
                        klog.Warning(err)
			continue
		}

Reply:
nice catch, yes, it only run once 😂